source/extensions/transformers/aws_lambda/api_gateway_transformer.cc

```diff
@@ -132,17 +134,32 @@ void ApiGatewayTransformer::transform_response(
   if (json_body.contains("multiValueHeaders")) {
     const auto& multi_value_headers = json_body["multiValueHeaders"];
     if (!multi_value_headers.is_object()) {
-        ENVOY_STREAM_LOG(debug, "invalid multi headers object", stream_filter_callbacks);
-        ApiGatewayError error = {500, "500", "invalid multi headers object"};
+        ENVOY_STREAM_LOG(debug, "invalid multiValueHeaders object", stream_filter_callbacks);
+        ApiGatewayError error = {500, "500", "invalid multiValueHeaders object"}; // TODO: update error message
         return ApiGatewayTransformer::format_error(*response_headers, body, error, stream_filter_callbacks);
     }
-
     for (json::const_iterator it = multi_value_headers.cbegin(); it != multi_value_headers.cend(); it++) {
         const auto& header_key = it.key();
         const auto& header_values = it.value();
 
+        // need to validate that header_values is an array/iterable
+        if (!header_values.is_array()) {
+          // if it's an object, we reject the request
+          if (header_values.is_object()) {
+            ENVOY_STREAM_LOG(debug, "invalid multi header value object", stream_filter_callbacks);
+            ApiGatewayError error = {500, "500", "invalid multi header value object"};
+            return ApiGatewayTransformer::format_error(*response_headers, body, error, stream_filter_callbacks);
+          }
+
+          ENVOY_STREAM_LOG(debug, "warning: using non-array value for multi header value", stream_filter_callbacks);
+        }
         for (json::const_iterator inner_it = header_values.cbegin(); inner_it != header_values.cend(); inner_it++) {
-          const auto& header_value = inner_it.value().get<std::string>();
+          std::string header_value;
+          if (inner_it.value().is_string()) {
+            header_value = inner_it.value().get<std::string>();
+          } else {
+            header_value = inner_it.value().dump();
```

Comment:
can we remove this conditional and just use .dump() -- does .dump() change string formatting?

Reply:
unfortunately this will not work for our purposes -- the "dumped" string value has escaped quotes, e.g. `\"test-header-value\"`

the documentation for `dump()` is [here](https://json.nlohmann.me/api/basic_json/dump/). From what I can tell, there isn't a setting that we can pass to `dump()` to work around this